src/main/java/com/ibm/cloudant/kafka/transforms/KeyToDocId.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright © 2022 IBM Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
+ * except in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+package com.ibm.cloudant.kafka.transforms;
+
+import org.apache.kafka.common.cache.Cache;
+import org.apache.kafka.common.cache.LRUCache;
+import org.apache.kafka.common.cache.SynchronizedCache;
+import org.apache.kafka.common.config.ConfigDef;
+import org.apache.kafka.connect.connector.ConnectRecord;
+import org.apache.kafka.connect.data.Field;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.SchemaBuilder;
+import org.apache.kafka.connect.data.Struct;
+import org.apache.kafka.connect.errors.DataException;
+import org.apache.kafka.connect.transforms.Transformation;
+import org.apache.kafka.connect.transforms.util.SchemaUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.apache.kafka.connect.transforms.util.Requirements.*;
+
+public class KeyToDocId<R extends ConnectRecord<R>> implements Transformation<R> {
+
+    public static final String OVERVIEW_DOC = "Extract the record's key and save the value to the '_id' field. " +
+            "If the '_id' record already exists, the value will be overwritten." +
+            "If the record key doesn't exist, the record will be unmodified.";
+
+    public static final ConfigDef CONFIG_DEF = new ConfigDef();
+
+    private static final String PURPOSE = "Copy record key to _id value";
+
+    private static final String idFieldName = "_id";
+
+    private Cache<Schema, Schema> schemaUpdateCache = new SynchronizedCache<>(new LRUCache<Schema, Schema>(16));
+
+    private static Logger LOG = LoggerFactory.getLogger(KeyToDocId.class.toString());
+
+    public void configure(Map<String, ?> configs) {
+        // no config required
+    }
+
+    public R apply(R record) {
+        if (record.key() != null
```

Comment:
If we use `util#requireSinkRecord` and it's a source record then a `DataException` will occur which puts Kafka in an unstable state.  What if we add:
```suggestion
        if (record.key() != null && record instanceof SinkRecord
```
and update the log message in the `else` block

Reply:
>and it's a source record then a DataException

Considering `_id` as a Cloudant specific requirement when we have a `SourceRecord` it will have come from Cloudant and already have an `_id`.
Adding an `_id` is only a requirement if you don't want generated IDs when writing and an `_id` didn't already exist in a record, so it is only relevant for our sink connector and hence `SinkRecord`.

As such I think a `DataException` is fine if it gets a `SourceRecord`.